### DISCUSSION TITLE
DisplayData convertion didn't work as expected

### DISCUSSION THREAD

**Discussion by ThomasUllrich666** [prometheus/snmp_exporter#994] [Q&A] [answered]
### Host operating system: output of `uname -a`
Linux sekidp-snmp-exporter-756d7b848c-5qj2h 4.18.0-477.13.1.el8_8.x86_64 #1 SMP Thu May 18 10:27:05 EDT 2023 x86_64 GNU/Linux

### snmp_exporter version: output of `snmp_exporter -version`
snmp_exporter, version 0.23.0 (branch: HEAD, revision: 20657d2f486a1ac237238827beed97fa67eb8611)
  build user:       root@f2310b8dd92c
  build date:       20230816-09:56:45
  go version:       go1.20.7
  platform:         linux/amd64
  tags:             netgo

### What device/snmpwalk OID are you using?
ThinkSystem SR630 V2
OID: 1.3.6.1.4.1.19046.11.1.1.1.2.1.3
https://www.circitor.fr/Mibs/Html/L/LENOVO-XCC-MIB.php#tempReading

### If this is a new device, please link to the MIB(s).
https://www.circitor.fr/Mibs/Html/L/LENOVO-XCC-MIB.php

### What did you do that produced an error?
I want to override the Display string parameter of the temperature value to a gauge datatype.
Currently the temperature value is represented as a String that lead to the situation that each value will be represented as a label in prometheus.  

Therefore I added the overrides parameter to mark it as gauge. 

Unfortunately this didn't work. The snmp agent captures always 0 instead of the values.

lenovo-xcc:
    walk:
      - temperature
      - voltage
      - fans
      - systemHealth
      - leds
      - hostMACAddressTable
      - xccVpdEntry
      - spTxtId
      - machineVpd
    overrides:
      voltReading:
        type: gauge
      voltNominalReading:
        type: gauge
      voltNonRecovLimitHigh:
        type: gauge
      voltNonCritLimitHigh:
        type: gauge
      voltNonRecovLimitLow:
        type: gauge
      voltCritLimitHigh:
        type: gauge
      voltCritLimitLow:
        type: gauge       
      voltNonCritLimitLow:
        type: gauge
      tempReading:
        type: gauge
      tempNominalReading:
        type: gauge
      tempNonRecovLimitHigh:
        type: gauge
      tempNonCritLimitHigh:
        type: gauge
      tempNonRecovLimitLow:
        type: gauge
      tempCritLimitLow:
        type: gauge
      tempNonCritLimitLow:
        type: gauge

### What did you expect to see?
the temperature values

### What did you see instead?
just 0

**Comment by x53414D** [comment]
I have the same issue, changing the DisplayString to gauge doesn't work.
Issue Overview

When monitoring Lenovo devices using the SNMP protocol, the values fetched are primarily in the STRING data type. This is inconsistent with other manufacturers, such as Dell, which provides more usable SNMP data types like INTEGER or GAUGE. 

MIB File in Use

lnvgy_fw_xcc_d8bt54m-5.00_lenovoxcc-anyos_noarch.mib

Example Data

Lenovo Devices

Most of the values from Lenovo devices are STRING:

SNMPv2-SMI::enterprises.19046.11.1.1.1.2.1.3.1 = STRING: "25"
SNMPv2-SMI::enterprises.19046.11.1.1.1.2.1.3.2 = STRING: "39"
...

Dell Devices

In contrast, values from Dell devices are usually INTEGER:

SNMPv2-SMI::enterprises.674.10892.5.4.700.20.1.6.1.1 = INTEGER: 620
SNMPv2-SMI::enterprises.674.10892.5.4.700.20.1.6.1.2 = INTEGER: 660
...

Specific Areas of Concern

Fan Speed Metrics: Currently provided in percentage and as a string.

SNMPv2-SMI::enterprises.19046.11.1.1.3.2.1.3.1 = STRING: " 74% of maximum"
...

 Desired Format: Values should be in RPM for effective graphing.

Fan Status: Also represented as strings.

SNMPv2-SMI::enterprises.19046.11.1.1.3.2.1.10.1 = STRING: "Normal"
...

Power Status: Presented as strings.

SNMPv2-SMI::enterprises.19046.11.1.1.11.2.1.6.1 = STRING: "Normal"
...

Conclusion

Having SNMP data in more quantifiable types like INTEGER or GAUGE would make monitoring and graphing significantly more straightforward and accurate. Therefore, it's crucial to address these discrepancies for more efficient device management.

**Comment by x53414D** [comment]
For instance, when monitoring fan speed metrics from Lenovo devices, the SNMP data returned is as follows:

fanSpeed{ fanIndex="1", fanSpeed=" 74% of maximum", hostname="####", instance="####", job="####", prometheus="####", prometheus_replica="####", rack="####", region="####", site="####", snmp="####", team="####"} and then 1

**Comment by SuperQ** [comment, marked as answer]
This is what the [regex_extracts](https://github.com/prometheus/snmp_exporter/blob/main/generator/README.md) is used for. It allows you to parse the numbers out of arbitrary strings.

**Reply by x53414D** [reply]
Could you please give me an example, I read this but cannot find the correct template for it.
what is the correct regex for this value:
fanSpeed{ fanIndex="1", fanSpeed=" 74% of maximum", hostname="####", instance="####", job="####", prometheus="####", prometheus_replica="####", rack="####", region="####", site="####", snmp="####", team="####"} and then 1

I want to have the 74 as a string value and the end.

**Reply by x53414D** [reply]
I added this but cannot fetch fanSpeed anymore in Prometheus (no value).
```
    overrides:
      fanSpeed:
        regex_extracts:
          fanSpeedtest:
            - regex: '(.*)'
              value: '$1'
```

**Comment by x53414D** [comment]
This is the final regex for fan speed, it returns the number (50) instead of (50%), but this created another metric called "**fanSpeedPercentage**"
 
```
    overrides:
# Extract fan speed percentage from string
      fanSpeed:
        type: DisplayString
        regex_extracts:
          Percentage:
            - regex: '\s*(\d{1,3}).*'
              value: '$1'
```
Thanks to Jacob from Lenovo team to clarify this matter.
